Add security section to FIP-0019 (#220)
filecoin-project/FIPs: FIPS/fip-0019.md

### Loosing epsilon-Replication guarantees

We work under the assumption that for each bit grinded on the {\rhoi}i=1,...,N  an adversary can flip one bit of her choice on the original R.

We consider \lambda = 80 as the security parameter. This means that we give the adversary a bit flip budget of 80 bits on R (by crafting R* such that these bits are flipped, keeping all the rest untouched) in order to make R* more compressible than R. 

> Observation: As we will see in what follows, in the case of our actual encoding (where a single \rhoi is shared among K distinct nodes Dj) we'll be either more conservative than we should, according to Kolmogorov bounds.

Author: It is D<sub>j</sub> which is a node of data.

#### Analysis 1: Security against "2 bytes for 1 bit" Adversary 

Preliminaries:
Given a predetermined string of L bits, we take for granted that a random string has probability 2<sup>-L</sup> to contain it.
We will consider bit strings of k consecutive 0s (or 1s). Note that it is the same when considering patterns like 01 or 10, the only difference is that those sequences would allow for less compression.
Assumption 2: An adversary can compress ~2 consecutive bytes (we would consider 2 bytes per node, but it can happen anywhere) by flipping a single bit
This means having substrings s that are of the form  s = 0000000100000000 (or its negation)
Note that the probability of s (or negation of s) to happen is 2<sup>-15</sup>. 
Since the adversary can adaptively choose the bits to flip, we are just assuming he finds 80 of these strings in a 32GiB string. 2 bytes out of 32 represent the 6.25% of a single node

Note 1: Remember that each randomness \rhoi is shared among K nodes, have 2<sup>30</sup>/K different hashes.
As a consequence, we consider each compression performed on a node D to propagate on all the K nodes that share the same randomness with D.

> Assumption 3: An adversary A is able to find a string of the form of s = 000000010000000 (or equivalent) in 80 different nodes, each of those uses a different \rhoi

Reviewer: ```suggestion
**Assumption 3**: An adversary A is able to find a string of the form of `s = 000000010000000` (or equivalent) in 80 different nodes, each of those uses a different `\rhoi`
```

> Note 2: Assumption 3 maximises the propagation of the compression made by the adversary, since all the local compression propagate to K nodes each. This results in a total compression of 80*0.0625*K nodes out of 2<sup>30</sup>

Reviewer: ```suggestion
Note 2: Assumption 3 maximizes the propagation of the compression made by the adversary since all the local compression propagates to K nodes each. This results in a total compression of 80*0.0625*K nodes out of 2<sup>30</sup>
```

> With H = 2<sup>9</sup> and K = 2<sup>21</sup> we obtain a total compression under 1%.

Reviewer: ```suggestion
With` H = 2<sup>9</sup>` and `K = 2<sup>21</sup>` we obtain a total compression under 1%.
```

Author: sup won't work in the code tags

> Assume we have 2<sup>9</sup> different hashes. This means that we have a total compression, according with Assumption 2 and Assumption 3 of ~1%

Reviewer: ```suggestion
Assume we have 2<sup>9</sup> different hashes. This means that we have a total compression, according to Assumption 2 and Assumption 3 of ~1%.
```

#### Analysis 2: Based on Kolmogorov Complexity (Information Theoretic Argument)
Information Theoretic Argument: Given a string S of size N with incompressibility V, for any  adversary that modifies S into S' by flipping T bits of his choice, we have that the incompressibility of S' is at least V - T*(log(N) + log(log(N))+1).
Proof:  Assume by contradiction that there exists a string S'' which is more compressible that S'. 
Then it means that it is possible to compress S into a string S* of length  | S'' | + T*(log(N) + 2*log(log(N))+1) < V 

> Note: T*(log(N) + 2*log(log(N))+1) are the bits needed in order to be able to store the T bit positions where the adversary flipped the bits, using self delimiting codes (References: this [paper](https://eprint.iacr.org/2021/162.pdf), page 9; this [book](https://core.ac.uk/download/pdf/301633813.pdf), Section 2.2)

Reviewer: ```suggestion
Note: _T*(log(N) + 2*log(log(N))+1)_ are the bits needed in order to be able to store the T bit positions where the adversary flipped the bits, using self delimiting codes (References: this [paper](https://eprint.iacr.org/2021/162.pdf), page 9; this [book](https://core.ac.uk/download/pdf/301633813.pdf), Section 2.2).
```

Given the following:

> We assume R to be incompressible, and we then set the incompressibility of R to N. 

Reviewer: ```suggestion
We assume `R` to be incompressible, and we then set the incompressibility of `R` to `N`. 
```

We assume an adversary can flip 80 bits of his choice in R
To be conservative, given that we are considering 2<sup>30</sup>/K different hashes, we apply the compressibility lower bound on a vector of size N' = 2<sup>30</sup>/K nodes

According to the lower bound, we get that an adversary that can flip 80 bits of his choice on R can not compress R to a string which has compressibility smaller than N' - 80(log(N') + 2*log(log(N')) + 1). 

> In order to be even more conservative, we are using N instead of N' in the compressibility reduction factor. We are then considering that R can not be compressed to a string which is less compressible than 2<sup>30</sup>/K  - 80(log(2<sup>30</sup>) + log(log(2<sup>30</sup>)) + 1) = 2<sup>30</sup>/ K - 80(30+11) 

Reviewer: ```suggestion
In order to be even more conservative, we are using `N` instead of `N'` in the compressibility reduction factor. We are then considering that R can not be compressed to a string which is less compressible than 2<sup>30</sup>/K  - 80(log(2<sup>30</sup>) + log(log(2<sup>30</sup>)) + 1) = 2<sup>30</sup>/ K - 80(30+11) .
```
